store.go

```diff
@@ -413,15 +414,32 @@ func (ls *LimitStore) Collect(id SpanID, anns ...Annotation) error {
 	ls.mu.Lock()
 	if ls.ring == nil {
 		ls.ring = make([]int64, ls.Max)
+		ls.traces = make(map[ID]struct{}, ls.Max)
 	}
+
+	// Check if the trace already exists in the ring. Otherwise, we would evict
+	// an old trace upon each annotation collection, rather than upon each new
+	// trace.
+	ids := make([]ID, 0)
+	for _, x := range ls.ring {
+		ids = append(ids, ID(x))
+	}
+	if _, ok := ls.traces[id.Trace]; ok {
+		ls.mu.Unlock()
+		return ls.DeleteStore.Collect(id, anns...)
+	}
+
 	if nextInsert := ls.ring[ls.nextInsertIdx]; nextInsert != 0 {
```

Comment:
That if statement only handles deletion of the old item in the ring. We don't want to run lines 437-439:

```
	ls.traces[id.Trace] = struct{}{}
	ls.ring[ls.nextInsertIdx] = int64(id.Trace)
	ls.nextInsertIdx = (ls.nextInsertIdx + 1) % ls.Max // increment & wrap
```

So to do it your way we'd have to write it as:

```
if _, exists := ls.traces[id.Trace}; !exists {
	if nextInsert := ls.ring[ls.nextInsertIdx]; nextInsert != 0 {
		// Store is at capacity (we know this because the next insert
		// slot already contains trace); delete oldest.
		old := ID(ls.ring[ls.nextInsertIdx])
		delete(ls.traces, old)
		if err := ls.DeleteStore.Delete(old); err != nil {
			return err
		}
	}
	ls.traces[id.Trace] = struct{}{}
	ls.ring[ls.nextInsertIdx] = int64(id.Trace)
	ls.nextInsertIdx = (ls.nextInsertIdx + 1) % ls.Max // increment & wrap
}
```

Is that still preferable? (I normally try to avoid extra if-statement nesting when possible by using early returns)